drivers/adc/adc_silabs_iadc.c

```diff
@@ -0,0 +1,688 @@
+		{ 5, _IADC_CFG_OSRHS_HISPD32, _IADC_CFG_DIGAVG_AVG1,
+		 _IADC_SCANFIFOCFG_ALIGNMENT_RIGHT16, 16 }, /* 32x oversampling*/
+		{ 6, _IADC_CFG_OSRHS_HISPD64, _IADC_CFG_DIGAVG_AVG1,
+		 _IADC_SCANFIFOCFG_ALIGNMENT_RIGHT20, 17 }, /* 64x oversampling*/
+		{ 7, _IADC_CFG_OSRHS_HISPD64, _IADC_CFG_DIGAVG_AVG2,
+		 _IADC_SCANFIFOCFG_ALIGNMENT_RIGHT20, 18 }, /* 128x oversampling*/
+		{ 8, _IADC_CFG_OSRHS_HISPD64, _IADC_CFG_DIGAVG_AVG4,
+		 _IADC_SCANFIFOCFG_ALIGNMENT_RIGHT20, 19 }, /* 256x oversampling*/
+		{ 9, _IADC_CFG_OSRHS_HISPD64, _IADC_CFG_DIGAVG_AVG8,
+		 _IADC_SCANFIFOCFG_ALIGNMENT_RIGHT20, 20 }, /* 512x oversampling*/
+		{ 10, _IADC_CFG_OSRHS_HISPD64, _IADC_CFG_DIGAVG_AVG16,
+		 _IADC_SCANFIFOCFG_ALIGNMENT_RIGHT20, 21 }, /* 1024x oversampling*/
+	};
+
+	if (!sequence->oversampling) {
+		/* Default to 2x oversampling (as in the default config)*/
+		data->oversampling = _IADC_CFG_OSRHS_HISPD2;
+		if (!sequence->resolution) {
+			/* For compliance with adc_api test, a resolution of 0 is considered as an
+			 * error
+			 */
+			return -EINVAL;
+			/* Default to 12 bit resolution (as in the default config)*/
+			/* data->alignment = _IADC_SCANFIFOCFG_ALIGNMENT_RIGHT12;
+			 * data->resolution = 12;
+			 */
+		} else {
+			res = sequence->resolution;
+
+			switch (res) {
+			case 12:
+				data->alignment = _IADC_SCANFIFOCFG_ALIGNMENT_RIGHT12;
+				data->resolution = 12;
+				break;
+			case 16:
+				data->alignment = _IADC_SCANFIFOCFG_ALIGNMENT_RIGHT16;
+				data->resolution = 16;
+				break;
+			case 20:
+				data->alignment = _IADC_SCANFIFOCFG_ALIGNMENT_RIGHT20;
+				data->resolution = 20;
+				break;
+			default:
+				LOG_ERR("Unsupported resolution %d", res);
+				return -EINVAL;
+			}
+		}
+	} else {
+		ospl = sequence->oversampling;
+
+		/* Max general samples averaging is 2^oversampling = 1024 */
+		if (ospl < 0 || ospl > 10) {
+			LOG_ERR("Unsupported oversampling %d", ospl);
+			return -EINVAL;
+		}
+		if (ospl > 6 && IS_ENABLED(IADC_NO_DIGAVG)) {
+			LOG_ERR("Unsupported oversampling %d", ospl);
+			return -EINVAL;
+		}
+
+		data->oversampling = ospl_table[ospl].analog_oversampling;
+		data->digital_averaging = ospl_table[ospl].digital_averaging;
+
+		if (!sequence->resolution) {
+			data->alignment = ospl_table[ospl].alignment;
+			if (ospl == 0 || ospl == 1) {
+				data->resolution = 12;
+			} else if (ospl >= 2 && ospl <= 5) {
+				data->resolution = 16;
+			} else {
+				data->resolution = 20;
+			}
+		} else {
+			res = sequence->resolution;
+
+			switch (res) {
+			case 12:
+				data->alignment = _IADC_SCANFIFOCFG_ALIGNMENT_RIGHT12;
+				data->resolution = 12;
+				break;
+			case 16:
+				data->alignment = _IADC_SCANFIFOCFG_ALIGNMENT_RIGHT16;
+				data->resolution = 16;
+				break;
+			case 20:
+				data->alignment = _IADC_SCANFIFOCFG_ALIGNMENT_RIGHT20;
+				data->resolution = 20;
+				break;
+			default:
+				LOG_ERR("Unsupported resolution %d", res);
+				return -EINVAL;
+			}
```

Comment:
This function is rather long and this switch case appear twice in the function. Maybe we can factorize the code in one way or another?
I wonder if we could even leverage `ospl_table` to get `alignment` and `resolution`.